src/nvm/dkg_keys.rs

```diff
@@ -94,46 +97,42 @@ impl DkgKeys {
 
     #[inline(never)]
     #[allow(unused)]
-    pub fn set_slice(&self, mut index: usize, value: &[u8]) -> Result<(), AppSW> {
-        self.check_write_pos(index + value.len())?;
+    pub fn set_slice(&self, index: usize, value: &[u8]) -> Result<(), AppSW> {
+        let end_index = index + value.len();
+        self.check_write_pos(end_index - 1)?; // Check only the last position
 
         let mut updated_data: [u8; DKG_KEYS_MAX_SIZE] = unsafe { *DATA.get_mut().get_ref() };
-        for b in value.iter() {
-            updated_data[index] = *b;
-            index += 1;
-        }
+
+        updated_data[index..end_index].copy_from_slice(value);
+
         unsafe {
             DATA.get_mut().update(&updated_data);
         }
-
         self.is_valid_write()?;
         Ok(())
     }
 
     #[inline(never)]
     #[allow(unused)]
-    pub fn set_slice_with_len(&self, mut index: usize, value: &[u8]) -> Result<usize, AppSW> {
+    pub fn set_slice_with_len(&self, index: usize, value: &[u8]) -> Result<usize, AppSW> {
         let len = value.len();
-        self.check_write_pos(index + 2 + len)?;
+        let total_len = 2 + len; // 2 bytes for length + actual data length
+        self.check_write_pos(index + total_len)?;
```

Comment:
@neithanmo on the set_slice fn, we are decreasing the index we check by 1. Shouldn't we do the same here?

Reply:
What do you mean -1?
I miss unit testing here, we could have written some tests for all this logic only?

Reply:
I also think that this module and buffer share a common API which can be abstracted away, but this can be left for later

Reply:
i do not see any index decreasing in both set_slice functions